### PR TITLE
VS Code extension + plugin

### PR DESCRIPTION
**Pull request by diarreola**
In the extension folder, there are two folders `vsc-js`, that has the javascript vs code extension which is the `extension/vsc-js/extension.js ` file , and `vsc-ts`, that has the typescript vs code extension which is the `extension/vsc-ts/howdoi/src/extension.ts` file. I wasn't sure if you preferred the extension in js or ts.

 In addition, in the `ext-plugin` folder, the `plugin.ts` contains the first version of the plugin code, which works with the command line output rather than a json object which I hope to do once it's available. The plugin returns a 2D array where each element in the array is an array containing an the answer and link, `[[answer1, link1],[answer2, link2],...]`.

### CODE REVIEW

File: ext-plugin/plugin.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        textToBeModified = textToBeModified.replace(regexBegins, '');
+        let result:string[] = [textToBeModified, commentBegin, ''];
+        return result;
+    }
+    else {
+        return null;
+    }
+}
+
+function spliceArr(obj:string[], commentBegin:string, commentEnd:string) {
```
Comment (gleitz) [inline ext-plugin/plugin.ts:85]: It's hard to read exactly what this function is doing. It might be because your result structure looks like `[[answer1, link1],[answer2, link2],...]` but `obj` and `newArr` is also not descriptive so it's difficult to wrap my head around this. Some high level suggestions:

What if the underlying return type looked like:
```
interface HowdoiResult {
   answer: string
   link: string
}
```
Then you could pass around HowdoiResult[] as a return type and reference it will result.answer and result.link.

Also, if you added typescript return types to all your functions `modifyCommentedText`, `howdoiResult`, this one, etc then it will be easier to read and in your editor you'll know exactly what is being returned as you write the code.

Finally, spliceArr could have a more descriptive name. `addCommentToResult`? `blockCommentAnswer`?

File: ext-plugin/plugin.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        newArr[i][0] = commentBegin + newArr[i][0] + commentEnd;
+    }
+    return newArr;
+}
+
+function howdoiResult(resultArr:string[],userTxt:string, commentBegin:string, commentEnd:string) {
+    const newResult:string[][] = spliceArr(resultArr,commentBegin,commentEnd);
+    return newResult;
+}
+
```
Comment (gleitz) [inline ext-plugin/plugin.ts:99]: nit: extra lines here at the bottom

File: ext-plugin/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig.json to read more about this file */
```
Comment (gleitz) [inline ext-plugin/tsconfig.json:3]: Might as well take out all the commented lines in this file.

File: extension/vsc-js/.eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (gleitz) [inline extension/vsc-js/.eslintrc.json:1]: I like 2-space indents rather than tabs

File: extension/vsc-js/.gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+node_modules/
```
Comment (gleitz) [inline extension/vsc-js/.gitignore:1]: thx! 🙏

File: extension/vsc-js/.vscode/extensions.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
+	// See https://go.microsoft.com/fwlink/?LinkId=733558
```
Comment (gleitz) [inline extension/vsc-js/.vscode/extensions.json:2]: tabs in this file too

File: extension/vsc-js/.vscode/launch.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
+// A launch configuration that launches the extension inside a new window
+// Use IntelliSense to learn about possible attributes.
+// Hover to view descriptions of existing attributes.
+// For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+{
+	"version": "0.2.0",
```
Comment (gleitz) [inline extension/vsc-js/.vscode/launch.json:6]: tabs in this file should be 2-space indents

File: extension/vsc-js/README.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+#5 Choose an option
+
+#6 The answer will be display in the code editor along with the question and link.
+
+## Requirements
+
+Install howdoi on your machine before using the howdoi extension.
+
+### Installation
```
Comment (gleitz) [inline extension/vsc-js/README.md:36]: Will this README appear on the extension's page? Perhaps link back to the main howdoi github page for installation instructions since those might change from time to time.

File: extension/vsc-js/extension.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
+// The module 'vscode' contains the VS Code extensibility API
```
Comment (gleitz) [inline extension/vsc-js/extension.js:1]: Was this file auto-generated from the Typescript file? Generally in Typescript projects I check in the .ts files but not the .js files. If we go "full-typescript" is this file still needed?

File: extension/vsc-js/jsconfig.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
+{
+	"compilerOptions": {
```
Comment (gleitz) [inline extension/vsc-js/jsconfig.json:2]: remove tabs

File: extension/vsc-js/package.json [hunk 11]
```diff
@@ -0,0 +1,40 @@
+{
+	"name": "howdoi",
```
Comment (gleitz) [inline extension/vsc-js/package.json:2]: There's some tabs in this file. Also should we call this `howdoi-vscode` or something like that?

File: extension/vsc-js/test/runTest.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
+const path = require('path');
+
+const { runTests } = require('vscode-test');
+
+async function main() {
+	try {
+		// The folder containing the Extension Manifest package.json
```
Comment (gleitz) [inline extension/vsc-js/test/runTest.js:7]: tabs here

File: extension/vsc-js/test/suite/extension.test.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+const assert = require('assert');
+
+// You can import and use all API from the 'vscode' module
+// as well as import your extension to test it
+const vscode = require('vscode');
+// const myExtension = require('../extension');
+
+suite('Extension Test Suite', () => {
+	vscode.window.showInformationMessage('Start all tests.');
```
Comment (gleitz) [inline extension/vsc-js/test/suite/extension.test.js:9]: perhaps we could remove the tests for now?

File: extension/vsc-js/test/suite/index.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
+const path = require('path');
```
Comment (gleitz) [inline extension/vsc-js/test/suite/index.js:1]: tabs here

File: extension/vsc-ts/howdoi/.vscode/extensions.json [hunk 15]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (gleitz) [inline extension/vsc-ts/howdoi/.vscode/extensions.json:1]: tabs here

File: extension/vsc-ts/howdoi/.vscode/launch.json [hunk 16]
```diff
@@ -0,0 +1,36 @@
+// A launch configuration that compiles the extension and then opens it inside a new window
+// Use IntelliSense to learn about possible attributes.
+// Hover to view descriptions of existing attributes.
+// For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+{
+	"version": "0.2.0",
+	"configurations": [
+		{
```
Comment (gleitz) [inline extension/vsc-ts/howdoi/.vscode/launch.json:8]: tabs here

File: extension/vsc-ts/howdoi/.vscode/tasks.json [hunk 17]
```diff
@@ -0,0 +1,20 @@
+// See https://go.microsoft.com/fwlink/?LinkId=733558
+// for the documentation about the tasks.json format
+{
+	"version": "2.0.0",
```
Comment (gleitz) [inline extension/vsc-ts/howdoi/.vscode/tasks.json:4]: tabs here

File: extension/vsc-ts/howdoi/README.md [hunk 18]
```diff
@@ -0,0 +1,65 @@
+# howdoi README
```
Comment (gleitz) [inline extension/vsc-ts/howdoi/README.md:1]: This readme looks not quite finished. If we decide to go with the `vsc-ts` version then let's make sure the README is up to date

File: extension/vsc-ts/howdoi/out/extension.js [hunk 19]
```diff
@@ -0,0 +1,126 @@
+"use strict";
```
Comment (gleitz) [inline extension/vsc-ts/howdoi/out/extension.js:1]: When it comes to these `out` files, can they be ignored from git?

File: extension/vsc-ts/howdoi/src/extension.ts [hunk 20]
```diff
@@ -0,0 +1,136 @@
+import * as vscode from 'vscode';
```
Comment (gleitz) [inline extension/vsc-ts/howdoi/src/extension.ts:1]: This seems to be a copy of the file from above. Is it possible to just have one version?

File: extension/vsc-ts/howdoi/src/test/runTest.ts [hunk 21]
```diff
@@ -0,0 +1,23 @@
+import * as path from 'path';
```
Comment (gleitz) [inline extension/vsc-ts/howdoi/src/test/runTest.ts:1]: tabs here

File: extension/vsc-ts/howdoi/tsconfig.json [hunk 22]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (gleitz) [inline extension/vsc-ts/howdoi/tsconfig.json:1]: tabs

File: pyvenv.cfg [hunk 23]
```diff
@@ -0,0 +1,3 @@
+home = C:\Users\GWCLT23\AppData\Local\Programs\Python\Python38-32
```
Comment (gleitz) [inline pyvenv.cfg:1]: Might want to ignore this file

File: extension/vsc-js/README.md (hunk 8 above)
Comment (diarreola) [reply extension/vsc-js/README.md:36]: yes this will appear on the extensions's page. Will link it back to the main howdoi GitHub page for installation instructions.

File: ext-plugin/plugin.ts [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+        console.log(`child process exited with code ${code}`);
+        callback(result);
+    });
+}
+
+// removes howdoi from command
+function howdoiPrefix(command:string) {
+    const prefix = "howdoi";
+    
+    if (command.includes(prefix)) {
```
[the remark this answers is not in the document]
Comment (diarreola) [reply ext-plugin/plugin.ts:44]: very true! thanks for the catch

File: extension/vsc-js/test/suite/extension.test.js (hunk 13 above)
Comment (diarreola) [reply extension/vsc-js/test/suite/extension.test.js:9]: these are auto-generated test files, should I still remove them?

File: extension/vsc-js/test/suite/index.js (hunk 14 above)
Comment (diarreola) [reply extension/vsc-js/test/suite/index.js:1]: these are auto-generated files, should I still remove the tabs?

File: extension/vsc-js/test/suite/extension.test.js (hunk 13 above)
Comment (gleitz) [reply extension/vsc-js/test/suite/extension.test.js:9]: Perhaps there is a basic test we can perform? If that's difficult, then I would remove altogether, or keep the file but add a comment that tests are not implemented yet.

File: extension/vsc-js/test/suite/index.js (hunk 14 above)
Comment (gleitz) [reply extension/vsc-js/test/suite/index.js:1]: Yes please
